File: lambda-api/UniRpc/GroupAuthorizations.ts
import { __PolicySets } from './GroupAuthorizationPolicies';

interface IGroupPolicy {
    Name: string;
    Services: {[service: string]: string | string[]};
}

interface IGroupPolicySet {
    [group: string]: IGroupPolicy;
}

interface IPolicySets {
    [policy: string]: IGroupPolicySet;
}

export function GroupAuthorizations(policy: string, group: string, service: string, method: string): boolean {
    if (policy == '*' && group == '*') return true;
    if (!__PolicySets[policy]) return false;
    let groupPolicySet: IGroupPolicySet = __PolicySets[policy];
    if (!groupPolicySet[group]) return false;
    let groupPolicy: IGroupPolicy = groupPolicySet[group];
    if (!groupPolicy[service]) return false;
    let servicePolicy: string | string[] = groupPolicy[service];
    if (typeof servicePolicy == 'string' && servicePolicy == '*') {
        return true;
    }
    if (Array.isArray(servicePolicy) && servicePolicy.includes(method)) {
        return true;
    }
    return false;
}

export function GroupClausesAuthorize(groups: string, service: string, method: string) {
    if (typeof groups != 'string') return false;
    for (let groupClause of groups.split(',')) {
        let sections = groupClause.split('.');
        let policy = sections[0], group = sections[1];
        if (GroupAuthorizations(policy, group, service, method)) return true;
    }
    return false;
}

File: tests/transformers/python/websocket-lambda-service/GroupAuthorizations.py
from typing import TypedDict, Union, List, Dict
from GroupAuthorizationPolicies import __PolicySets


class IGroupPolicy(TypedDict):
    Name: str
    Services: Dict[str, Union[str, List[str]]]


IGroupPolicySet = Dict[str, IGroupPolicy]

IPolicySets = Dict[str, IGroupPolicySet]


def GroupAuthorizations(policy: str, group: str, service: str, method: str) -> bool:
    if policy == '*' and group == '*':
        return True
    if policy not in __PolicySets.keys():
        return False
    groupPolicySet: IGroupPolicySet = __PolicySets[policy]
    if group not in groupPolicySet.keys():
        return False
    groupPolicy: IGroupPolicy = groupPolicySet[group]
    if service not in groupPolicy.keys():
        return False
    servicePolicy: Union[str, List[str]] = groupPolicy[service]
    if isinstance(servicePolicy, str) and servicePolicy == '*':
        return True
    if isinstance(servicePolicy, list) and servicePolicy.includes(method):
        return True
    return False


def GroupClausesAuthorize(groups: str, service: str, method: str) -> bool:
    if not isinstance(groups, str):
        return False
    for groupClause in groups.split(','):
        sections = groupClause.split('.')
        policy = sections[0]
        group = sections[1]
        if GroupAuthorizations(policy, group, service, method):
            return True
    return False
